src/targets/gpu/include/migraphx/gpu/gemm.hpp

```diff
@@ -10,21 +13,196 @@ namespace gpu {
 
 struct context;
 
-struct miopen_gemm
+template <class Op>
+struct rocblas_gemm
 {
-    op::dot op;
+    Op op;
 
     template <class Self, class F>
     static auto reflect(Self& self, F f)
     {
         return migraphx::reflect(self.op, f);
     }
 
-    std::string name() const { return "gpu::gemm"; }
-    shape compute_shape(const std::vector<shape>& inputs) const;
+    std::string name() const
+    {
+        if(op.name().find("quant_") != std::string::npos)
+        {
+            return "gpu::quant_gemm";
+        }
+        else
+        {
+            return "gpu::gemm";
+        }
+    }
+
+    shape compute_shape(const std::vector<shape>& inputs) const
+    {
+        std::vector<shape> in_shapes(inputs);
+        in_shapes.pop_back();
+        check_shapes{in_shapes}.not_broadcasted();
+        batch_not_transposed(inputs[0].strides());
+        batch_not_transposed(inputs[1].strides());
+
+        return op.compute_shape(in_shapes);
+    }
+
+    rocblas_datatype get_type(shape::type_t type) const
+    {
+        switch(type)
+        {
+        case shape::double_type: return rocblas_datatype_f64_r;
+        case shape::float_type: return rocblas_datatype_f32_r;
+        case shape::half_type: return rocblas_datatype_f16_r;
+        case shape::int8_type: return rocblas_datatype_i8_r;
+        case shape::uint8_type: return rocblas_datatype_u8_r;
+        case shape::int32_type: return rocblas_datatype_i32_r;
+        case shape::uint32_type: return rocblas_datatype_u32_r;
+        case shape::uint16_type:
+        case shape::int16_type:
+        case shape::int64_type:
+        case shape::uint64_type: MIGRAPHX_THROW("ROCBLAS_GEMM: data type not supported!");
+        }
+
+        MIGRAPHX_THROW("ROCBLAS_GEMM: data type not supported!");
+
+        return rocblas_datatype_f32_r;
+    }
+
     argument
-    compute(context& ctx, const shape& output_shape, const std::vector<argument>& args) const;
-    void batch_not_transposed(const std::vector<std::size_t>& strides) const;
+    compute(context& ctx, const shape& output_shape, const std::vector<argument>& args) const
+    {
+        bool transa     = args[0].get_shape().transposed();
```

Comment:
I would prefer the compute method to go into a cpp file, or at least put this into a function that is implemented in the cpp file.